pandas/tests/frame/test_nonunique_indexes.py

```diff
@@ -513,3 +513,9 @@ def test_set_value_by_index(self):
 
         df.iloc[:, 0] = 3
         tm.assert_series_equal(df.iloc[:, 1], expected)
+
+    def test_masking_duplicate_columns(self):
+        # https://github.com/pandas-dev/pandas/issues/31954
+        df = pd.DataFrame(np.arange(4).reshape(2, 2), columns=[0, 0])
+        expected = DataFrame([[np.nan, np.nan], [np.nan, 3.0]], columns=[0, 0])
+        tm.assert_frame_equal(df[df > 2], expected)
```

Comment:
Hmm, so this actually did not raise before the bug was fixed. I think a minimal example for this would be something like a DataFrame with both float and int dtype columns but with the same name. From the commit just before 1c0cc62e30a3077476e97f8e7e6ba17b4ac754b6:

```python
[ins] In [47]: df
Out[47]: 
   0  0
0  0  1
1  2  3

[ins] In [48]: df[df > 2]
Out[48]: 
    0    0
0 NaN  NaN
1 NaN  3.0

...

[ins] In [56]: df
Out[56]: 
     0  0
0  0.0  1
1  2.0  3

[ins] In [57]: df[df > 2]  # this raises a ValueError

```